x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java

```diff
@@ -596,6 +630,270 @@ public void testByteFieldWithIntSubfieldTooBig() throws IOException {
+                index.startObject("file");
+                {
+                    index.field("type", "object");
+                    index.startObject("properties");
+                    {
+                        index.startObject("raw").field("type", "keyword").endObject();
+                    }
+                    index.endObject();
+                }
+                index.endObject();
+            }
+            index.endObject();
+        });
+        index("test2", """
+            {"file": {"raw": "o2"}}""");
+
+        ResponseException e = expectThrows(
+            ResponseException.class,
+            () -> runEsqlSync(new RestEsqlTestCase.RequestObjectBuilder().query("FROM test* | SORT file, file.raw | LIMIT 3"))
+        );
+        String err = EntityUtils.toString(e.getResponse().getEntity());
+        assertThat(err, containsString("mapped as [2] incompatible types: [keyword] in [test1], [object] in [test2]"));
+
+        Map<String, Object> result = runEsqlSync(new RestEsqlTestCase.RequestObjectBuilder().query("FROM test* | SORT file.raw | LIMIT 2"));
+        assertMap(
+            result,
+            matchesMap().entry("columns", List.of(columnInfo("file", "unsupported"), columnInfo("file.raw", "keyword")))
+                .entry("values", List.of(matchesList().item(null).item("o2"), matchesList().item(null).item(null)))
+        );
+    }
+
+    /**
+     * One index with an unsupported field and a supported sub-field. The supported sub-field
+     * is marked as unsupported <strong>because</strong> the parent is unsupported. Mapping:
+     * <pre>
+     * "f": {
+     *    "type": "ip_range"  ----- The type here doesn't matter, but it has to be one we don't support
+     *    "fields": {
+     *       "raw": {
+     *          "type": "keyword"
+     *       }
+     *    }
+     * }
+     * </pre>.
+     */
+    public void testPropagateUnsupportedToSubFields() throws IOException {
+        createIndex("test", index -> {
+            index.startObject("properties");
+            index.startObject("f");
+            {
+                index.field("type", "ip_range");
+                index.startObject("fields");
+                {
+                    index.startObject("raw").field("type", "keyword").endObject();
+                }
+                index.endObject();
+            }
+            index.endObject();
+            index.endObject();
+        });
+        index("test", """
+            {"f": "192.168.0.1/24"}""");
+
+        ResponseException e = expectThrows(
+            ResponseException.class,
+            () -> runEsqlSync(new RestEsqlTestCase.RequestObjectBuilder().query("FROM test* | SORT f, f.raw | LIMIT 3"))
+        );
+        String err = EntityUtils.toString(e.getResponse().getEntity());
+        assertThat(err, containsString("Cannot use field [f] with unsupported type [ip_range]"));
+        assertThat(err, containsString("Cannot use field [f.raw] with unsupported type [ip_range]"));
```

Comment:
I don't remember why we marked the hierarchy of unsupported fields as unsupported as well. My best quess is that, since we "materialized" the values of subfields from the parent fields (because this is what actually gets indexed as part of the `_source`), and parents had types that couldn't be displayed by SQL, we set the entire hierarchy as unsupported. For the record, [this PR](https://github.com/elastic/elasticsearch/pull/50823/files) added this behavior. @costin do you remember why we did this for SQL? Is my guess above correct?

Nevertheless, is this "limitation" still valid for ESQL who reads the values from Lucene differently than what it used to be for SQL?

For example, this use case does seem to make sense for ESQL, unless I am forgetting/missing something:

test1

```
            "file": {
                "type": "ip_range",
                "fields": {
                    "raw": {
                        "type": "keyword"
                    }
                }
            }
```

test2

```
            "file": {
                "type": "text",
                "fields": {
                    "raw": {
                        "type": "keyword"
                    }
                }
            }
```

For query `from test*` we return `file.raw` as `unsupported`.

Nik's test above replicates what happens in `main`, but `main` seems to be broken from at least one angle: `file.raw` is not of type `ip_range` as the error message says.

--- 

### In conclusion:
- should hierarchies of unsupported field types be also marked as unsupported (current behavior in SQL and old IndexResolver)? If so, why?
- there is some context related to this specific use case:
  - for SQL we did this change https://github.com/elastic/elasticsearch/pull/50823/files but **it may be only SQL specific**
  - we already changed the SQL's default behavior slightly with this PR https://github.com/elastic/elasticsearch/pull/100875/files where we recognized that ESQL is different from SQL in one matter

Reply:
Great points @astefan . Tl;dr is that ESQL already has and will have better handling of field extraction and thus conflict resolution than SQL and by having a different resolved, we should be able to evolve this better.
in SQL extraction is done through field caps hence why we expected a whole field hierarchy to match - it might not have been perfect but it got us pretty far.

In ESQL that assumption no longer holds - #100875 allowed subfields to allow the following [mapping](https://github.com/elastic/elasticsearch/pull/100875/files#diff-70d0893691f1bd1e3b4be7b7146fbbbb1168851e8cdcf9ace6c26e5b6839b1a9):

index 1 - `field.name` keyword
index 2 - `field` = keyword

In Index 1, field is object, as parent for field.name. This conflicts with field in index2 which is a keyword - using the SQL behavior we discarded field.name as unsupported. However there's no `field.name` in index2 and thus there's no mapping conflict - in other words, field is problematic but field.name is not.

I think we're in a good enough place with this and considering the performance savings I suggest we move forward with it - there's a chance there might be some subtle cases we haven't covered yet and by changing the behavior we're going to run into them but right now we have enough testing in place to be confident about them.
Furthermore I would backport this behavior in 8.13 to test it as much as possible. Potentially introduce an internal flag to have the fallback behavior for a while when running into problems.